driver-sync/src/test/functional/com/mongodb/client/RetryableWritesProseTest.java

```diff
@@ -0,0 +1,205 @@
+
+// See https://github.com/mongodb/specifications/tree/master/source/retryable-writes/tests/README.rst#replica-set-failover-test
+public class RetryableWritesProseTest extends DatabaseTestCase {
+    private MongoClient clientUnderTest;
+    private MongoClient failPointClient;
+    private MongoClient stepDownClient;
+    private final String databaseName;
+    private final String collectionName;
+    private final String description;
+
+    public RetryableWritesProseTest() {
+        this.description = "Replica set failover test for retryable writes";
+        this.databaseName = getDefaultDatabaseName();
+        this.collectionName = "test";
+    }
+
+    @BeforeClass
+    public static void beforeClass() {
+    }
+
+    @AfterClass
+    public static void afterClass() {
+    }
+
+    @Before
+    @Override
+    public void setUp() {
+        assumeTrue(canRunTests());
+
+        setUpClientUnderTest();
+        setUpFailPointClient();
+        setUpStepDownClient();
+    }
+
+    @After
+    public void cleanUp() {
+        if (failPointClient != null) {
+            MongoDatabase failPointAdminDB = failPointClient.getDatabase("admin");
+            failPointAdminDB.runCommand(new Document("mode", "off"));
+        }
+
+        if (clientUnderTest != null) {
+            clientUnderTest.close();
+        }
+        if (stepDownClient != null) {
+            stepDownClient.close();
+        }
+    }
+
+    @Test
+    public void shouldPassAllOutcomes() {
+        // Using the client under test, insert a document and observe a successful write result.
+        // This will ensure that initial discovery takes place.
+        MongoDatabase database = clientUnderTest.getDatabase(databaseName);
+        MongoCollection<Document> collection = database.getCollection(collectionName);
+        Document doc = new Document("_id", 1).append("x", 11);
+        try {
+            System.out.println("--- inserting initial document...");
+            collection.insertOne(doc);
+            System.out.println("--- DONE inserting initial document");
+        } catch (MongoServerException ex) {
+            fail("Initial insert failed");
+        }
+
+        // Using the fail point client, activate the fail point by setting mode to "alwaysOn".
+        activateFailPoint();
+
+        // Using the step down client, step down the primary by executing the command { replSetStepDown: 60, force: true}.
+        // This operation will hang so long as the fail point is activated. When the fail point is later deactivated,
+        // the step down will complete and the primary's client connections will be dropped. At that point, any ensuing
+        // network error should be ignored.
+        MongoDatabase stepDownDB = stepDownPrimary();
+
+
+        System.out.println("Waiting for secondary...");
+        try {
+            while (!stepDownDB.runCommand(new BsonDocument("isMaster", new BsonInt32(1))).getBoolean("secondary")) {
```

Comment:
Here, I wait for the primary to become a secondary before attempting the next insert.